File: 338. Counting Bits.py
class Solution:
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        ans = [0]
        m = 1
        for i in range(1, num + 1):
            if i >= 2 * m:
                m *= 2
            ans.append(ans[i - m] + 1)
        return ans


sol = Solution()
print(sol.countBits(0))
print(sol.countBits(1))
print(sol.countBits(5))

File: 338. Counting Bits.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> ans(num+1, 0);
        for (int i=1; i <= num; ++i) {
            ans[i] = ans[i&(i-1)] + 1;
        }
        return ans;
    }
};

int main() {
    auto ans = Solution().countBits(10);
    for (int i: ans) cout << i << ' ';
    cout << endl;
    return 0;
}
